atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -0,0 +1,189 @@
+
+    def get_excluded_platforms(self):
+        df_dir = self.workflow.source.get_dockerfile_path()[1]
+        exclude_platforms = set()
+        exclude_path = os.path.join(df_dir, self.EXCLUDE_FILENAME)
+        if os.path.exists(exclude_path):
+            with open(exclude_path) as f:
+                for platform in f:
+                    platform = platform.strip()
+                    if not platform:
+                        continue
+                    exclude_platforms.add(platform)
+
+        return exclude_platforms
+
+    def get_platforms(self):
+        return self.platforms - self.get_excluded_platforms()
+
+    def get_current_builds(self, osbs):
+        field_selector = ','.join(['status!={status}'.format(status=status.capitalize())
+                                   for status in BUILD_FINISHED_STATES])
+        return len(osbs.list_builds(field_selector=field_selector))
+
+    def get_cluster_info(self, cluster, platform):
+        osbs = OSBS(Configuration(conf_section=cluster.name),
+                    Configuration(conf_section=cluster.name))
+        current_builds = self.get_current_builds(osbs)
+        load = current_builds / cluster.max_concurrent_builds
+        return ClusterInfo(cluster, platform, osbs, load)
+
+    def choose_cluster(self, platform):
+        config = get_config(self.workflow)
+        clusters = [self.get_cluster_info(cluster, platform) for cluster in
+                    config.get_enabled_clusters_for_platform(platform)]
+
+        if not clusters:
+            raise RuntimeError('No clusters found for platform {}!'
+                               .format(platform))
+
+        selected = min(clusters, key=lambda c: c.load)
+        self.log.info('platform %s will use cluster %s',
+                      platform, selected.cluster.name)
+        return selected
+
+    def get_release(self):
+        labels = df_parser(self.workflow.builder.df_path, workflow=self.workflow).labels
+        return get_preferred_label(labels, 'release')
+
+    def get_worker_build_kwargs(self, release, platform):
+        build_kwargs = deepcopy(self.build_kwargs)
+
+        build_kwargs.pop('architecture', None)
+
+        build_kwargs['release'] = release
+        build_kwargs['platform'] = platform
+
+        return build_kwargs
+
+    def run(self):
+        release = self.get_release()
+        worker_builds = []
+
+        clusters = [self.choose_cluster(platform)
+                    for platform in self.get_platforms()]
+
+        for cluster_info in clusters:
+            kwargs = self.get_worker_build_kwargs(release, cluster_info.platform)
+            worker_build = cluster_info.osbs.create_worker_build(**kwargs)
+            worker_builds.append(
+                WorkerBuildInfo(build=worker_build, cluster_info=cluster_info))
+
+        for build_info in worker_builds:
+            build_info.wait_to_schedule()
+
+        thread_pool = ThreadPool(len(worker_builds))
```

Comment:
Modified so that the life cycle of each worker build is handled by its own thread. This has the benefit of avoid an overall failure if one of the builds fails to be monitored properly.

Using `map_async` with short wait intervals to ensure build cancellation is handled immediately. Verified that when atomic_reactor receives SIGTERM, worker builds are canceled properly.